### PR TITLE
feat: add datasource for Hermit package manager

### PR DESCRIPTION
**Pull request by lyonlai**
## Changes

As part of the Hermit manager feature from #16250, this PR contains the datasource component used by Hermit package manager.

## Context

Issue: #16119
PR: #16250

## Documentation (please check one with an [x])

- [X] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [X] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/hermit/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.spec.ts:3]:
```suggestion
import { HermitDatasource } from '.';
```

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:15]: inline this

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:139]: inline this small function

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:148]: log a debug message and return null. we prefer to not throw errors from datasource

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:89]: move this validation to `getReleases`.

log an error and return null.

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:98]: see other comment. do we really need custom registry support?

**Review by viceice — CHANGES_REQUESTED** [review]
don't throw errors from datasource. log appropriate messages and return null instead

File: lib/modules/datasource/hermit/index.ts
Comment (lyonlai) [reply lib/modules/datasource/hermit/index.ts:99]: We do need custom registry support. We in Square have our own `hermit-packages` containing internal packages that won't be available in the open sourced `cashapp/hermit-packages`.

File: lib/modules/datasource/hermit/index.ts
Comment (PhilipAbed) [inline lib/modules/datasource/hermit/index.ts:122]: can we simplify this by regex?

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:36]:
```suggestion
  http: GithubHttp;
```
override base property

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:58]: i think we can error here, as it should never happen at runtime

@rarkins WDYT?


```suggestion
      logger.error('registryUrl must be supplied');
```

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:141]: what happens for normal `http.getJson` ? 🤔 is GitHub returning wrong content type for JSON files?

File: lib/modules/datasource/github-releases/types.ts
Comment (viceice) [inline lib/modules/datasource/github-releases/types.ts:16]: why this? can't we reuse the browser download url as in other datasource/ manager? 🤔

Comment (lyonlai) [reply lib/modules/datasource/github-releases/types.ts:16]: nah. `browser_download_url` doesn't work with private repository assets. I've tested it

File: lib/modules/datasource/hermit/index.ts
Comment (lyonlai) [reply lib/modules/datasource/hermit/index.ts:141]: https://github.com/renovatebot/renovate/pull/16255#discussion_r907136376

**Comment by lyonlai** [comment]
👋 @viceice  & @PhilipAbed  the build is green now after the merge of Hermit versioning. Can you please take a look again?

**Comment by lyonlai** [comment]
@viceice & @PhilipAbed  any update on this? any chance to get this merge today?

File: lib/modules/datasource/hermit/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.spec.ts:41]:
```suggestion
      httpMock.scope(githubApiHost).get(indexAssetUrl).reply(200, resp);
```
no need to stringify before

File: lib/modules/datasource/hermit/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.spec.ts:24]:
```suggestion
```
see below

File: lib/modules/datasource/hermit/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.spec.ts:96]:
```suggestion
      httpMock.scope(githubApiHost).get(indexAssetUrl).reply(200, []);
```

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:74]:
```suggestion
      logger.debug({ packageName, registryUrl }, 'cannot find hermit package');
```
this isn't an error. we usually don't fail here. logging an error would cause renovate to exit with error.

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:135]: did you tried to use `getJson` function?

Comment (lyonlai) [reply lib/modules/datasource/hermit/index.ts:135]: The very first implementation I use `http.get` to get the search manifest. But you didn't like it back then and told me to use `http.stream` like the examples from `terraform` and `bazel`. I did and I made the change on #16255 so it works with the current implementation. Can we please be consistent on the same topic?

https://github.com/renovatebot/renovate/pull/16255#discussion_r907136376

Comment (PhilipAbed) [reply lib/modules/datasource/hermit/index.ts:135]: not sure it might happen here but i've seen this happen before, where URL doesnt return JSON but returns 404 in an HTML error
there's no handling for the `JSON.parse(indexContent)` here in case it fails parsing or returns some HTML instead of a json..

so i think to use what @viceice mentioned above is better

Comment (lyonlai) [reply lib/modules/datasource/hermit/index.ts:135]: @viceice was very not happy and asked to use `http.stream` instead (see [this comment](https://github.com/renovatebot/renovate/pull/16255#discussion_r907136376)), which I did.  I would like not to make any undesired changes unless a clear and acceptable reason given.

Comment (viceice) [reply lib/modules/datasource/hermit/index.ts:135]: It was because before it wasn't clear to me, that we download `JSON` here. So why we need the `accept` header?
Shouldn't the default `getJson` work? which should send a `accept: application/json` and return the already parsed string as object.

Comment (lyonlai) [reply lib/modules/datasource/hermit/index.ts:135]: We need to use the `accept` header `application/octet-stream` because it will be needed to download Release Asset for private Github Repository. See [Github Release Asset Document](https://docs.github.com/en/rest/releases/assets).  

I have [explained](https://github.com/renovatebot/renovate/pull/16257#discussion_r914470336) before in our case we are downloading release assets from private Github Repository instead of the open source one. 

Hitting the URL with `getJSON` won't work because hitting the URL with `getJSON` will only return the content of the API call, not the content of the manifest.  like the example screenshot from Github's document.

image

Comment (PhilipAbed) [reply lib/modules/datasource/hermit/index.ts:135]: if you need the manifest then just handle the JSON parsing with try/catch or throw error depending on your logic and it should be ok i guess

Comment (lyonlai) [reply lib/modules/datasource/hermit/index.ts:135]: Alright, this sounds much more reasonable. I'll wrap the `JSON.parse` with try catch and return `null` instead, which fits the logic in the invoking function.  Thanks @PhilipAbed

Comment (viceice) [reply lib/modules/datasource/hermit/index.ts:135]: and also please add some code comment, why you need to use `accept` header and streaming here

Comment (lyonlai) [reply lib/modules/datasource/hermit/index.ts:142]: @PhilipAbed this is done as discussed, PTAL.

**Comment by lyonlai** [comment]
@viceice & @PhilipAbed any chance to get this merged today? Changes are done as per discussed/suggested.

**Review by PhilipAbed — APPROVED** [review]

**Comment by PhilipAbed** [comment]
i approved, lets wait for the maintainers to approve

**Review by PhilipAbed — DISMISSED** [review]

**Comment by lyonlai** [comment]
@rarkins Can you jump in and give me a hand for the merge? This is taking way too long.

**Comment by rarkins** [comment]
@lyonlai it's not ok to keep complaining about how quickly people respond to you in an open source project. There's already been 10s of hours spent helping your PRs get ready for merge. If you think this approach so far is not OK, then close your PRs and keep hermit in a private fork.

I am on vacation, so it will be up to the other maintainers.

**Comment by lyonlai** [comment]
I appreciate the help so far on getting this PR ready to merge. But I really want to get the feature in before I get pulled off to other duties.

Sorry to bother you on your holiday @rarkins, didn't know about this. I'll wait for other maintainers to respond.

File: lib/modules/datasource/github-releases/types.ts
Comment (viceice) [inline lib/modules/datasource/github-releases/types.ts:17]: We probably need to use `url` in other datasources too? otherwise it will fail for private repos as @lyonlai  suggested?

File: lib/modules/datasource/hermit/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.spec.ts:15]:
```suggestion
      const resp: HermitSearchResult[] = [
```

File: lib/modules/datasource/hermit/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.spec.ts:23]:
```suggestion
      ];
```

this case will hide type errors

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:40]: i think this can cause issues, if a user add's ad hostrule for `hermit` instead of `github-releases`. We should do like we do for `pod` datasource
```suggestion
    this.http = new GithubHttp(id);
```

https://github.com/renovatebot/renovate/blob/5fa2aa2f2c0de8eccba46acfde84934a239c4b4b/lib/constants/platforms.ts#L11-L17
https://github.com/renovatebot/renovate/blob/5fa2aa2f2c0de8eccba46acfde84934a239c4b4b/lib/constants/platform.spec.ts#L45

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:106]:
```suggestion
    const u = parseUrl(registryUrl);
```

use `parseUrl` from `utils/url`

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:150]:
```suggestion
```

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:152]:
```suggestion
      return JSON.parse(indexContent) as HermitSearchResult[];
```

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:157]:
```suggestion
    return null;
```

File: lib/modules/datasource/github-releases/types.ts
Comment (viceice) [reply lib/modules/datasource/github-releases/types.ts:17]: @HonkingGoose please file an issue for this 🤗

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:18]:
```suggestion
 * Hermit Datasource searches a given package from the specified `hermit-packages`
```

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:21]: i think we should remove that sentence as it's not really right

File: lib/modules/datasource/github-releases/types.ts
Comment (HonkingGoose) [reply lib/modules/datasource/github-releases/types.ts:17]: Done, see:

- #16575

File: lib/modules/datasource/hermit/readme.md
Comment (HonkingGoose) [inline lib/modules/datasource/hermit/readme.md:1]:
```suggestion
By default [Hermit](https://cashapp.github.io/hermit/) looks up packages from the open source project [https://github.com/cashapp/hermit-packages](https://github.com/cashapp/hermit-packages).
```

File: lib/modules/datasource/hermit/readme.md
Comment (HonkingGoose) [inline lib/modules/datasource/hermit/readme.md:3]:
```suggestion
Hermit supports [private packages](https://cashapp.github.io/hermit/packaging/private/).
To get Renovate to find your private packages, follow these steps:
```

File: lib/modules/datasource/hermit/readme.md
Comment (HonkingGoose) [inline lib/modules/datasource/hermit/readme.md:7]:
```suggestion
1. perform `hermit search --json` with your private Hermit distribution and save the file to `index.json`
1. make a GitHub release in your private packages repository named `index` with the asset `index.json` generated in step 1.
1. setup a CI pipeline to repeat step 1 & 2 on new commits to the private packages repository.
```

- The rendered Markdown numbers each item for us. We can just use `1.` to create a numbered list.
- GitHub is a proper noun, so should be capitalized.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Some initial changes.

**Comment by HonkingGoose** [comment]
@lyonlai Next time, please "batch" the commit suggestions so you're not committing multiple times in a row, this causes a lot of notifications.

Read the [GitHub docs, incorporating feedback in your pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request) to learn how to batch all the suggestions into a single commit.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
One final change, please review this PR:

- https://github.com/lyonlai/renovate/pull/1

File: lib/modules/datasource/hermit/readme.md
Comment (lyonlai) [reply lib/modules/datasource/hermit/readme.md:8]: > I guess you need to follow these steps in your package's repository? It looks like you create a index.json file, and then publish it in a GitHub release?

Yes, it will be needed to be done in package's repository. like [cashapp/hermit-packages](https://github.com/cashapp/hermit-packages).

GitHub action is one way to do the publish. That is easily achievable in many other CI pipelines. That's why I write down the steps but not really outline it in a specific platform.

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:72]: pass `parsedUrl`, so you don't need to parse again

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:112]: you don't need to pass a object, simply pass the parsed url object

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:119]: maybe log a warn message about wrong registry url?

File: lib/modules/datasource/hermit/index.ts
Comment (viceice) [inline lib/modules/datasource/hermit/index.ts:132]:
```suggestion
      (asset) => asset.name === 'index.json'
```
typescript should be able to resolve the type

File: lib/modules/datasource/hermit/index.ts
Comment (lyonlai) [reply lib/modules/datasource/hermit/index.ts:72]: I did pass it in a previous commit. however yesterday when I was trying it. the `key` function for cache in `getHermitSearchManifest` gets an empty object instead. where I couldn't turn it into a string like I do now.  It causes error when I run Renovate against our environment. 

Not sure it Is a known issue with the `cache` decorator.?

Comment (lyonlai) [reply lib/modules/datasource/hermit/index.ts]: tried it again. the empty object seems to relate to the logger filtering. key function now properly returns string of the given parsed url object.  might did something wrong yesterday. never mind

**Review by viceice — APPROVED** [review]
🎉

**Pull request merged**
